### PR TITLE
gh-117151: IO performance improvement, increase io.DEFAULT_BUFFER_SIZE to 128k

### PR DESCRIPTION
**Pull request by morotti**
See discussion gh-117151

This patch adjusts the buffer size. That gives 3 to 5 times I/O performance improvement on modern hardware.

* increase io.DEFAULT_BUFFER_SIZE to 128k
* fix open() to use max(st_blksize, io.DEFAULT_BUFFER_SIZE)

### CODE REVIEW

**Comment by morotti** [comment]
@serhiy-storchaka hello, you were kind to review my first PR for buffering performance (118037), would you be able to review this PR too?

File: Lib/_pyio.py [hunk 1]
```diff
@@ -23,8 +23,8 @@
     valid_seek_flags.add(os.SEEK_HOLE)
     valid_seek_flags.add(os.SEEK_DATA)
 
-# open() uses st_blksize whenever we can
-DEFAULT_BUFFER_SIZE = 8 * 1024  # bytes
+# open() uses max(st_blksize, io.DEFAULT_BUFFER_SIZE) when st_blksize is available
+DEFAULT_BUFFER_SIZE = 128 * 1024  # bytes
```
Comment (eendebakpt) [inline Lib/_pyio.py:28]: This value is used for the python only implementation of `io`. Performance considerations could be different for the python implementation, but I think it is good to indeed update this to the same value as in `_iomodule.h`.

**Comment by eendebakpt** [comment]
Pinging @benjaminp as expert on io. Could you review this PR? Are there other benchmarks (besides the one in the corresponding issue) we can perform to test this PR?

**Comment by morotti** [comment]
Thank you for reviewing,

The CLA check is green now and I added a news entry.

**Comment by morotti** [comment]
@eendebakpt @benjaminp could you review?

**Comment by eendebakpt** [comment]
> @eendebakpt @benjaminp could you review?

@morotti The PR looks good from my side, but I am no core dev so I cannot approve. Currently many core devs are at pycon US, so I think we should wait a bit more. If in a few weeks there has been no further response, we can post a message to discourse (see https://devguide.python.org/getting-started/pull-request-lifecycle/#reviewing).

**Comment by erlend-aasland** [comment]
Requesting Serhiy's review, since he reviewd the other linked (and merged) PR #118037.

File: Misc/NEWS.d/next/Core and Builtins/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+increase io.DEFAULT_BUFFER_SIZE from 4k/8k to 128k. fix open on Linux to use
+max(io.DEFAULT_BUFFER_SIZE, device block size) rather than the device block
```
Comment (eendebakpt) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst]:
```suggestion
``max(io.DEFAULT_BUFFER_SIZE, device block size)`` rather than the device block
```

File: Misc/NEWS.d/next/Core and Builtins/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst [hunk 3]
```diff
@@ -0,0 +1,3 @@
...
+size. improve I/O performance by 3 to 5 times. patch by Romain Morotti.
```
Comment (eendebakpt) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst]:
```suggestion
size. Improve I/O performance upto 3 to 5 times. Patch by Romain Morotti.
```

File: Misc/NEWS.d/next/Core and Builtins/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+increase io.DEFAULT_BUFFER_SIZE from 4k/8k to 128k. fix open on Linux to use
```
Comment (eendebakpt) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst]:
```suggestion
Increase ``io.DEFAULT_BUFFER_SIZE`` from 4k/8k to 128k. Adapt :func:`open` on Linux to use
```

File: Lib/_pyio.py [hunk 5]
```diff
@@ -249,7 +249,7 @@ def open(file, mode="r", buffering=-1, encoding=None, errors=None,
                 pass
             else:
                 if bs > 1:
```
Comment (serhiy-storchaka) [inline Lib/_pyio.py]: Isn't this redundant if use the maximal value? This can be either
```py
if bs > buffering:
    buffering = bs
```
or
```py
buffering = max(bs, buffering)
```
(buffering is set to DEFAULT_BUFFER_SIZE above).

BTW, it looks like the C code always uses the block size, even if it is smaller than DEFAULT_BUFFER_SIZE. So the NEWS entry is not correct.

File: Misc/NEWS.d/next/Core and Builtins/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst [hunk 6]
```diff
@@ -0,0 +1,3 @@
+Increase ``io.DEFAULT_BUFFER_SIZE`` from 4k/8k to 128k. Adapt :func:`open` on Linux to use
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst]: I think that this should be rather in the "Library" section than in "Core and Builtins".

**Review by serhiy-storchaka — COMMENTED** [review]
I was convinced that such a change could be beneficial.

But for the case if it has unexpected effect, please ask on https://discuss.python.org/. Update also the C implementation of `open()` which is used by default.

File: Misc/NEWS.d/next/Core and Builtins/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst (hunk 6 above)
Comment (morotti) [reply Misc/NEWS.d/next/Core and Builtins/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst]: This was set here by the news tool automatically. I think "Core and Builtins" makes more sense for the I/O buffering code to be considered core and builtins, though I can move if you prefer?

File: Lib/_pyio.py (hunk 5 above)
Comment (morotti) [reply Lib/_pyio.py]: I am gonna have a look at the C code and triple check what it does and what is should do.
I think open() used the block size when available (4k on Linux usually) or default 8k. The C buffering code used 8k by default and did not seem to rely on the block size. 
The python API sets the buffer size arguments when invoking the C classes, so it works as expected, but C code invoking the C API would not have the behavior (should it? it doesn't need to because the C buffering code operates on non-files too and doesn't have the same performance consideration as the python code, but I will have a look).

**Comment by serhiy-storchaka** [comment]
Please do not use rebase and force-push. It makes reviewing more difficult.

File: Lib/_pyio.py (hunk 5 above)
Comment (serhiy-storchaka) [reply Lib/_pyio.py]: Add a test that opens a large binary file with the default buffering, call `read1()` or `peek()`, and check the size of the result. It should be the same for both implementations.

File: Misc/NEWS.d/next/Core and Builtins/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst (hunk 6 above)
Comment (serhiy-storchaka) [reply Misc/NEWS.d/next/Core and Builtins/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst]: While `open()` is exposed as a builtin, it is defined in the `io` module. This is not a part of the core interpreter.

**Comment by morotti** [comment]
apologies, I thought I had to squash the new commits and force push.

**Comment by morotti** [comment]
all adjusted

* found and corrected another code path in the C API that was using the block size
* added a test
* updated the documentation

**Comment by cmaloney** [comment]
@morotti 
1. winconsoleio picks this up (https://github.com/python/cpython/blob/main/Modules/_io/winconsoleio.c#L425), and has a past bug that there is a cap (https://bugs.python.org/issue11395, https://github.com/python/cpython/blob/main/Modules/_io/winconsoleio.c#L1027-L1031). I don't think this PR breaks that (winconsoleio in `write()` cuts up too big of writes), but likely the default buffer size there shold be separated from this new bigger constant which is always going to require splitting write calls.
1. do you have any performance numbers around local disk `read()` with this change? (Curious for small and large local cached file)

**Comment by morotti** [comment]
thanks, having a look,

looking at the old threads, the bug with winconsole destroying large input was reported in 2010.
it was Windows XP era (2002-2014), it's not claer if the bug still applies to active windows versions. the limitation is no longer listed in the MSDN page.
support for windows XP was dropped forever ago, support for windows 7 was dropped with the python 3.11 release in 2022.

> https://tahoe-lafs.org/trac/tahoe-lafs/ticket/1232
> Correction: the [​MSDN documentation for WriteConsoleW](http://msdn.microsoft.com/en-us/library/ms687401%28VS.85%29.aspx) does document a length limitation, but it says 64 KiB (which would be 32767 UTF-16 characters and a terminating NUL), not 26608 characters. The fix is the same regardless; I'll limit it to 10000 [UTF-16] characters.

**Comment by morotti** [comment]
@cmaloney 

I've adjusted winconsole to have a different default buffer size.
I've added comments with what I could gather from the threads.


for the impact on read(), it's quite variable.

* larger `write()` gives 3 to 5 times performance improvements, just from the I/O.
* larger `read()` gives barely a few percent improvements (less than 10% on my machines), just from the I/O. the filesystem is really saving us from this terrible code! 

HOWEVER, that's just for the I/O. the app will do something with the chunks and any python code that is doing not-so-small operation on each chunk can get a very large impact.

we've found the bug in pip twice, https://github.com/pypa/pip/pull/12810
there is code to download packages in 10240 bytes chunks + there is code to read package files and compute the checksum in io.DEFAULT_BUFFER_SIZE=8192 chunks.

the download code is obliterated by the small chunks. it's calling some code and updating a progress bar with each chunk, as much as 30% of the pip runtime is updating the progress bar when it's downloading packages 🫨.
the hashing code is merely `hashlib.update(buffer)` in a loop. it's single digit percent improvement for using larger chunks. 
so in the same app, you get both a case where large chunks make an enormous difference and a not so much difference.

it's worthwhile to note that I've found python 3.11 to be 4 times faster than previous versions in tight loop and list comprehension. older python versions would get much more impact from small chunks, due to the larger overhead of many more loop iterations and function calls.

**Comment by cmaloney** [comment]
Soungs good to me. re: Windows, I dug into the docs and history a bit and it looks like with Windows 8+ it's not necessary anymore. I'm seeing if it's possible to remove the special case in gh-121940.

File: Misc/NEWS.d/next/Library/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst
Comment (gpshead) [inline Misc/NEWS.d/next/Library/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst:1]: FYI - Please check the box to always allow other maintainers to directly make edits to your PR branches.

Replace this NEWS file contents with:

```
Increase ``io.DEFAULT_BUFFER_SIZE`` from 8k to 128k and adjust :func:`open` on
platforms where ``fstat`` provides a ``st_blksize`` field (such as Linux) to use
``max(io.DEFAULT_BUFFER_SIZE, device block size)`` rather than always using the
device block size. This should improve I/O performance.
Patch by Romain Morotti.
```

File: Lib/_pyio.py [hunk 7]
```diff
@@ -1558,15 +1559,18 @@ def __init__(self, file, mode='r', closefd=True, opener=None):
                     os.set_inheritable(fd, False)
 
             self._closefd = closefd
-            self._stat_atopen = os.fstat(fd)
```
Comment (cmaloney) [inline Lib/_pyio.py]: This needs to merge changes with https://github.com/python/cpython/pull/123412 (the stat result is now stashed as a member so can do a  number of optimizations with it)

**Comment by gpshead** [comment]
overall I think this PR is good but it needs updating and consideration of the conflicts.

one caveat: instead of `max(blksize, DEFAULT_BUFFER_SIZE)` we should have a maximum `blksize` that we heed. If some filesystem returns a huge `st_blksize` (entirely out of our control), we don't want to blindly always attempt to allocate that much during IO.  `max(min(blksize, 4096*1024), DEFAULT_BUFFER_SIZE))` perhaps.

**Comment by cmaloney** [comment]
re: pyio, active project of mine to improve / bring to par with C FileIO :) gh-129005

**Comment by morotti** [comment]
ahah I saw, I was literally posting a comment in your discussion at the same second :D :D :D

File: Lib/_pyio.py [hunk 8]
```diff
@@ -123,10 +123,11 @@ def open(file, mode="r", buffering=-1, encoding=None, errors=None,
     the size of a fixed-size chunk buffer.  When no buffering argument is
     given, the default buffering policy works as follows:
 
-    * Binary files are buffered in fixed-size chunks; the size of the buffer
-      is chosen using a heuristic trying to determine the underlying device's
-      "block size" and falling back on `io.DEFAULT_BUFFER_SIZE`.
-      On many systems, the buffer will typically be 4096 or 8192 bytes long.
+   * Binary files are buffered in fixed-size chunks; the size of the buffer
+     is set to `max(io.DEFAULT_BUFFER_SIZE, st_blksize)` using a heuristic
```
Comment (cmaloney) [inline Lib/_pyio.py]: `st_blksize` here is used before it's ever defined, how about something like:

Binary files are buffered in fixed-size chunks; the size of the buffer is the max of the default block size defaults to the file's block size. On most systems, the buffer will typically be 131072 bytes long.

File: Lib/test/test_file.py [hunk 9]
```diff
@@ -216,6 +216,18 @@ def testSetBufferSize(self):
         with self.assertWarnsRegex(RuntimeWarning, 'line buffering'):
             self._checkBufferSize(1)
 
+    def testDefaultBufferSize(self):
+        f = self.open(TESTFN, 'wb')
+        blksize = f.raw._blksize
+        f.write(bytes([0] * 5_000_000))
```
Comment (cmaloney) [inline Lib/test/test_file.py]: f.write(b"\0" * 5_000_000)

Should be quite a bit faster to do the same thing (Avoids constructing a 5 million element list, then converting each element one at a time from integer to byte and copying across to the `bytes`.

File: Lib/test/test_file.py [hunk 10]
```diff
@@ -216,6 +216,18 @@ def testSetBufferSize(self):
         with self.assertWarnsRegex(RuntimeWarning, 'line buffering'):
             self._checkBufferSize(1)
 
+    def testDefaultBufferSize(self):
+        f = self.open(TESTFN, 'wb')
```
Comment (cmaloney) [inline Lib/test/test_file.py]: Should use `with` here. Definitely unlikely that `f.write()` will fail, but if it does, this doesn't close the file.

File: Misc/NEWS.d/next/Library/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst [hunk 11]
```diff
@@ -0,0 +1,5 @@
+Increase ``io.DEFAULT_BUFFER_SIZE`` from 8k to 128k and adjust :func:`open` on
+platforms where ``fstat`` provides a ``st_blksize`` field (such as Linux) to use
```
Comment (cmaloney) [inline Misc/NEWS.d/next/Library/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst]:
```
:meth:`os.fstat`
```

(Will make it link to those docs, which is what I think a reader of the news would want if they need more context)

File: Modules/_io/_iomodule.c [hunk 12]
```diff
@@ -132,9 +132,10 @@ the size of a fixed-size chunk buffer.  When no buffering argument is
 given, the default buffering policy works as follows:
 
 * Binary files are buffered in fixed-size chunks; the size of the buffer
-  is chosen using a heuristic trying to determine the underlying device's
-  "block size" and falling back on `io.DEFAULT_BUFFER_SIZE`.
-  On many systems, the buffer will typically be 4096 or 8192 bytes long.
+  is set to `max(io.DEFAULT_BUFFER_SIZE, st_blksize)` using a heuristic
```
Comment (cmaloney) [inline Modules/_io/_iomodule.c]: same not as `_pyio`, st_blksize is here before it's ever defined. I think don't need to mention it directly.

File: Modules/_io/_iomodule.c [hunk 13]
```diff
@@ -368,6 +369,8 @@ _io_open_impl(PyObject *module, PyObject *file, const char *mode,
         if (blksize_obj == NULL)
             goto error;
         buffering = PyLong_AsLong(blksize_obj);
+        if (buffering < DEFAULT_BUFFER_SIZE)
```
Comment (cmaloney) [inline Modules/_io/_iomodule.c]: All new `if` / code structure needs to do curly braces (the _io_open_impl code is fairly old, hence doesn't have a lot)

see: https://peps.python.org/pep-0007/

**Comment by gpshead** [comment]
You always want an in process maximum. Several megabytes is a reasonable
max due to memory architectures.

The underlying st_blksize is not something you can blindly trust and is
_not_ guidance for how best to do reads. The OS already does read-ahead
into its buffer cache for you.

File: Lib/test/test_file.py (hunk 10 above)
Comment (morotti) [reply Lib/test/test_file.py]: I think it's intentional to not use with? none of the tests in this file are using with. I think they don't expect the file class and the with context manager to be working.

Comment (cmaloney) [reply Lib/test/test_file.py]: Depends which test. There's some general "test suite" setup/teardown which covers a number of cases / prevents leaks, `test_fileio` uses with, I think it's good practice. Can definitely get "you're implementing a Python and some tests working", but this is new / beyond baseline "does open work".

a fileio test that uses with:
```python
    # verify readinto
    def testReadintoByteArray(self):
        self.f.write(bytes([1, 2, 0, 255]))
        self.f.close()

        ba = bytearray(b'abcdefgh')
        with self.FileIO(TESTFN, 'r') as f:
            n = f.readinto(ba)
        self.assertEqual(ba, b'\x01\x02\x00\xffefgh')
        self.assertEqual(n, 4)
```

**Comment by morotti** [comment]
alright, had to rebase on master since some build script changes just went in 1 hour ago.

I've added a 8MB limit as requested. That should be reasonable. I'd prefer not to do lower as there are use cases for multiple MB block sizes. (I don't even see the point of introducing a cap, when the cpython interpreter as been blindly trusting st_blksize for decades. Nobody ever complaint it was too large).

Now that begs the question, should the docstring be updated to say that the maximum of the minimum of the block size and 8MB and the default buffer size? ^^

File: Lib/test/test_file.py (hunk 10 above)
Comment (gpshead) [reply Lib/test/test_file.py]: stdlib code is old, don't assume existing code, especially in tests, necessarily sets a good example. :)

File: Modules/_io/_iomodule.c [hunk 14]
```diff
@@ -368,6 +367,14 @@ _io_open_impl(PyObject *module, PyObject *file, const char *mode,
         if (blksize_obj == NULL)
             goto error;
         buffering = PyLong_AsLong(blksize_obj);
+        if (buffering > 8192 * 1024)
```
Comment (eendebakpt) [inline Modules/_io/_iomodule.c]:
```suggestion
        if (buffering > MAXIMUM_BUFFER_SIZE)
```
and put `MAXIMUM_BUFFER_SIZE` in the header file next to `DEFAULT_BUFFER_SIZE`?

File: Lib/_pyio.py [hunk 15]
```diff
@@ -123,10 +123,9 @@ def open(file, mode="r", buffering=-1, encoding=None, errors=None,
     the size of a fixed-size chunk buffer.  When no buffering argument is
     given, the default buffering policy works as follows:
 
-    * Binary files are buffered in fixed-size chunks; the size of the buffer
-      is chosen using a heuristic trying to determine the underlying device's
-      "block size" and falling back on `io.DEFAULT_BUFFER_SIZE`.
-      On many systems, the buffer will typically be 4096 or 8192 bytes long.
+   * Binary files are buffered in fixed-size chunks; the size of the buffer
+     is the maximum of the DEFAULT_BUFFER_SIZE and the device block size.
+     On most systems, the buffer will typically be 131072 bytes long.
```
Comment (eendebakpt) [inline Lib/_pyio.py]:
```suggestion
     On most systems, the buffer will typically be 128 kilobytes long.
```
(or `128*1024`)

**Review by eendebakpt — COMMENTED** [review]
Two very minor comments (feel free to ignore).

The change itself (increasing the default buffer size) looks good to me!

File: Modules/_io/_iomodule.c (hunk 14 above)
Comment (cmaloney) [inline Modules/_io/_iomodule.c]: The curly brace style here isn't python standard

Moving to standard would be:
```C
if (buffering > 8192 * 1024) {
    buffering = 8192 * 1024;
}
if (buffering < DEFAULT_BUFFER_SIZE) {
    buffering = DEFAULT_BUFFER_SIZE;
}
```

although practically, i think Py_MIN and Py_MAX are going to be cleaner here:
`buffering = Py_MAX(Py_MIN(buffering, 8192 * 1024), DEFAULT_BUFFER_SIZE);`

**Comment by morotti** [comment]
alright, I think I've made all the adjustment requested :)

the new constant MAXIMUM_BUFFER_SIZE is not exposed to python.  do we want to expose it?
```
import io
io.MAXIMUM_BUFFER_SIZE
AttributeError: module 'io' has no attribute 'MAXIMUM_BUFFER_SIZE'
```

**Comment by morotti** [comment]
it appears that the github action run for `[Tests / Ubuntu (free-threading) / build and test (ubuntu-24.04-arm)` does not work
https://github.com/python/cpython/issues/129664

File: Modules/_io/_iomodule.c [hunk 16]
```diff
@@ -655,6 +660,10 @@ iomodule_exec(PyObject *m)
     if (PyModule_AddIntMacro(m, DEFAULT_BUFFER_SIZE) < 0)
         return -1;
 
+    /* MAXIMUM_BUFFER_SIZE */
+    if (PyModule_AddIntMacro(m, MAXIMUM_BUFFER_SIZE) < 0)
```
Comment (cmaloney) [inline Modules/_io/_iomodule.c]: I'm a bit torn on if this should be a new public Python API / constant. It feels like an implementation detail tuning point for optimizing.... This turns the new constant from implementation detail to public Python constant which should have at least one test and users may rely on in ways that are unexpected. 

If keep, need at least one new test around, although a simple one (check the type and value, probably in `test_io`?)

Comment (morotti) [reply Modules/_io/_iomodule.c]: I was torn too, which is why I didn't add it initially. ^^
Turns out, the tests are verifying that every constant is defined and exposed throughout the I/O APIs, so I had to expose it to pass.
I could try to make it a private variable `_MAXIMUM_BUFFER_SIZE` to see whether the tests ignore that?
Otherwise I can rename to `MAXIMUM_BLKSIZE` for clarity? I don't want people mislead into thinking there is a maximum buffer size. Computers have GB of memory and software should use the memory.

Comment (morotti) [reply Modules/_io/_iomodule.c]: Or maybe the better approach is to remove the maximum entirely? There is no maximum currently and it works well. There is no reason to expect block devices to be misconfigured to report block size of dozens of MB.

File: Lib/_pyio.py [hunk 17]
```diff
@@ -242,7 +244,7 @@ def open(file, mode="r", buffering=-1, encoding=None, errors=None,
             buffering = -1
             line_buffering = True
         if buffering < 0:
-            buffering = raw._blksize
```
Comment (cmaloney) [inline Lib/_pyio.py:245]: I'd make it named `_MAXIMUM_BUFFER_SIZE` (underscore leading should make it private), then removing the export from `_io` I think will pass tests.

I think the checks are 1. `_all_` contains everything without an underscore, 2. `_all_` matches between `_io` and `_pyio`

https://github.com/python/cpython/blob/main/Lib/test/test_io.py#L4216-L4219 which uses https://github.com/python/cpython/blob/main/Lib/test/support/__init__.py#L1627-L1683

File: Modules/_io/_iomodule.c (hunk 16 above)
Comment (cmaloney) [reply Modules/_io/_iomodule.c]: Worry without a max is weird platform behavior becomes python bugs, whereas with a max we "constrain" how out of ideal things can go. A little slower than "max speed" but no really bad cases is a lot better than "if you do this you get quadratic or exponential behavior" (aka your program seems to just hang infinitely, you need fairly deep knowledge of tools to figure / dive into why).

My napkin math: 256GB/s PICe max bandwidth bus today, with 128 MB buffer, would mean ~8000 round trips through python where defaulted buffer size logic is used in that second to "saturate". That's about the worst case / I don't think fully saturating sata with fixed-sized reads in python is likely; the "correctly estimated file size" is likely a better route...). That sort of case would almost certainly be better served by using `bytearray` and `os.readinto` / `recvfrom` which avoid temporary objects and copies for things that expose the [Buffer Protocol](https://docs.python.org/3/c-api/buffer.html#bufferobjects). That doesn't do this cap.

There's a lot of small devices that also run python installed in the world (ex. Home Assistant is written in Python and can run on Raspberry Pi 3 model B, which has 1GB of RAM). 128MB is a very large portion of that, and I can definitely see little home / custom / development devices exposing a bad st_blksize. Too big likely means paging (Linux is good at overallocating "untouched" memory generally, but still). Tiny memory limit containers also fall into this somewhat, and small container reading from big-blocksize device in server farm running kubernetes seems plausible. 

A third case is there are platforms and devices (ex. Pipes) where large reads _will always be capped smaller_ so we would over allocate then resize. Those ideally _should_ to me be using Python's I/O stack (although often do hand coded variants of I/O primitives today, but not always). Over-sizing there makes a significant performance impact: https://github.com/python/cpython/issues/121313

Comment (morotti) [reply Modules/_io/_iomodule.c]: I don't understand the reference to 128MB? the PR is setting the default buffer to 128 kB and capping the maximum to 8 MB.
Are you implying that there could be bad devices with a block size set as high as 128 MB. I can understand the concern but I don't think that that ever happened. The cpython intepreter had been using the block size as the buffer size for the past decade, without any limit. People would have realized the interpreter was broken (running out of memory), if their system worked at all. 
That being said, I think it's fine to have a limit, I just don't want it too low, because whole MB block size is really a thing, and if multiple MB may not common now, it is still possible and it will certainly be becoming more common over the coming decade as hardware continues to grow.

Comment (cmaloney) [reply Modules/_io/_iomodule.c]: Yea, I swapped up sizes in my head. 128Kb definitely a lot better for small platforms. 8MB also a lot better percentage of total memory.

Bad devices set as high as 128 MB, weird things (ex. https://github.com/python/cpython/issues/93287), corruption / hardware bugs resulting in weird values in the field, tuning max blocksize for particular places (NFS, etc). For things which really want to go fast, they probably don't want to do readall using this loop which always makes a new buffer from scratch.

Bigger it gets, feels like optimizing for Python on "big" systems at the expense of small. Already microcontrollers tend to use things like MicroPython and/or CircuitPython, be nice to not push more platforms to need alternate python implementations to run well (I like CPython covering as many use cases as possible "good enough")

File: Modules/_io/_iomodule.c [hunk 18]
```diff
@@ -60,8 +60,12 @@ PyDoc_STRVAR(module_doc,
 "DEFAULT_BUFFER_SIZE\n"
 "\n"
 "   An int containing the default buffer size used by the module's buffered\n"
-"   I/O classes. open() uses the file's blksize (as obtained by os.stat) if\n"
-"   possible.\n"
+"   I/O classes.\n"
+"\n"
+"_MAXIMUM_BUFFER_SIZE\n"
```
Comment (cmaloney) [inline Modules/_io/_iomodule.c]: this bit should go away I think?

File: Lib/_pyio.py [hunk 19]
```diff
@@ -123,10 +125,10 @@ def open(file, mode="r", buffering=-1, encoding=None, errors=None,
     the size of a fixed-size chunk buffer.  When no buffering argument is
     given, the default buffering policy works as follows:
 
-    * Binary files are buffered in fixed-size chunks; the size of the buffer
-      is chosen using a heuristic trying to determine the underlying device's
-      "block size" and falling back on `io.DEFAULT_BUFFER_SIZE`.
-      On many systems, the buffer will typically be 4096 or 8192 bytes long.
+   * Binary files are buffered in fixed-size chunks; the size of the buffer
+     is max(min(blocksize, _MAXIMUM_BUFFER_SIZE), DEFAULT_BUFFER_SIZE)
```
Comment (cmaloney) [inline Lib/_pyio.py]: Since we're not publicly exposing _MAXIMUM_BUFFER_SIZE, I think should use `8192 * 1024` directly.

File: Doc/library/functions.rst [hunk 20]
```diff
@@ -1405,10 +1405,10 @@ are always available.  They are listed here in alphabetical order.
    :func:`io.TextIOWrapper.reconfigure`. When no *buffering* argument is
    given, the default buffering policy works as follows:
 
-   * Binary files are buffered in fixed-size chunks; the size of the buffer is
-     chosen using a heuristic trying to determine the underlying device's "block
-     size" and falling back on :const:`io.DEFAULT_BUFFER_SIZE`.  On many systems,
-     the buffer will typically be 4096 or 8192 bytes long.
+   * Binary files are buffered in fixed-size chunks; the size of the buffer
+     is max(min(blocksize, _MAXIMUM_BUFFER_SIZE), DEFAULT_BUFFER_SIZE)
```
Comment (cmaloney) [inline Doc/library/functions.rst]: _MAXIMUM_BUFFER_SIZE -> 8192 * 1024

**Review by cmaloney — APPROVED** [review]
Generally looking good to me, just small tweaks

**Review by cmaloney — COMMENTED** [review]
It looks like the comment and bits excluding winconsoleio got lost in one of the rebases and the like?

**Comment by morotti** [comment]
I think you were looking into the winconsoleio optimization? 
you have a bug ticket here https://github.com/python/cpython/issues/121940
and commit here https://github.com/cmaloney/cpython/commit/7adcb0dba5e44e877b6cabb3301eb540888e214c

I think we should keep it as a separate issue for a separate PR, or we'll never complete anything. ^^

I do have an old branch with comments, from when I was investigating too https://github.com/man-group/cpython/blob/io-buffer-size-archive/Modules/_io/winconsoleio.c#L51 (The comments and the branch are obsolete as some other people have redone some of the buffering already, the 10 year old code was counting characters wrong or something).

TL;DR Old versions of windows were buggy with large writes to the console, theoretically above 65k, practically with as little as 26k, so there is code in winconsole to split writes into smaller writes. This is a windows XP era bug. Few people had a look since we reopened the bug and confirmed it's no longer an issue in windows 8. Old the code could be removed. (I think you have branches and PR to do that)

**Comment by morotti** [comment]
alright, I've removed the constants _MAXIMUM_BUFFER_SIZE. build green.

I think we're good to merge

**Review by cmaloney — APPROVED** [review]

**Comment by erlend-aasland** [comment]
> (rebasing on master again because CI builds no longer work)

Please use `git merge --no-ff main` instead of rebasing. This has already been pointed out in earlier review.

File: Modules/_io/_iomodule.c [hunk 21]
```diff
@@ -368,6 +367,7 @@ _io_open_impl(PyObject *module, PyObject *file, const char *mode,
         if (blksize_obj == NULL)
             goto error;
         buffering = PyLong_AsLong(blksize_obj);
+        buffering = Py_MAX(Py_MIN(buffering, 8192 * 1024), DEFAULT_BUFFER_SIZE);
```
Comment (gpshead) [inline Modules/_io/_iomodule.c]: This needs to go below the error check, otherwise the -1 return value PyErr could be ignored.

**Review by gpshead — CHANGES_REQUESTED** [review]

File: Misc/NEWS.d/next/Library/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst [hunk 22]
```diff
@@ -0,0 +1,5 @@
+Increase ``io.DEFAULT_BUFFER_SIZE`` from 8k to 128k and adjust :func:`open` on
+platforms where :meth:`os.fstat` provides a ``st_blksize`` field (such as Linux)
+to use ``max(min(io.DEFAULT_BUFFER_SIZE, 8 MiB), device block size)`` rather
```
Comment (gpshead) [inline Misc/NEWS.d/next/Library/2024-04-30-14-03-09.gh-issue-117151.yt2H8c.rst]: this doesn't match the code.

**Comment by morotti** [comment]
@gpshead I've made the changes, would you be ablet to review again?

**Comment by morotti** [comment]
hello @gpshead @cmaloney @serhiy-storchaka @eendebakpt  would you be able to review and unblock the PR?

**Review by gpshead — APPROVED** [review]

File: Doc/library/functions.rst [hunk 23]
```diff
@@ -1405,10 +1405,10 @@ are always available.  They are listed here in alphabetical order.
    :func:`io.TextIOWrapper.reconfigure`. When no *buffering* argument is
    given, the default buffering policy works as follows:
 
-   * Binary files are buffered in fixed-size chunks; the size of the buffer is
-     chosen using a heuristic trying to determine the underlying device's "block
-     size" and falling back on :const:`io.DEFAULT_BUFFER_SIZE`.  On many systems,
-     the buffer will typically be 4096 or 8192 bytes long.
+   * Binary files are buffered in fixed-size chunks; the size of the buffer
+     is max(min(blocksize, 8 MiB), DEFAULT_BUFFER_SIZE)
```
Comment (gpshead) [inline Doc/library/functions.rst]: surround the code in double backticks?

**Review by gpshead — COMMENTED** [review]
the change looks good, but this PR was created with the checkbox allowing committers to make direct edits to the PR branch disabled so we can't take care of trivia that may be blocking it ourselves.

can you merge master so that it reruns modern CI?

**Comment by morotti** [comment]
@gpshead merged with master.

sorry, I haven't found any checkbox to allow to make direct edits or I would have ticket. I suspect it's because the fork is in my work organization, maybe that comes with extra restrictions.
